engine/src/main/java/org/terasology/network/BroadCastServer.java

```diff
@@ -0,0 +1,139 @@
+            });
+        } catch (Exception e) {
+            logger.info("Broadcasting has been interrupted " + e.getMessage());
+            stopBroadCast();
+        }
+    }
+
+    public void stopBroadCast() {
+        BroadCastServer.turnOnBroadcast = false;
+        logger.info("Shutting down BroadCasting");
+        this.sendSocket.close();
+        this.receiveSocket.close();
+    }
+
+    private void listenToBroadCast() throws Exception {
+        byte[] buffer = DISCOVERY_REQUEST.getBytes(StandardCharsets.UTF_8);
+        DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
+        getReceiveSocket().receive(packet);
+        logger.info("Discovery package received! -> " + packet.getAddress() + ":" + packet.getPort());
+
+        //Validate the data sent.
+        String data = new String(packet.getData()).trim();
+        if (data.equals(DISCOVERY_REQUEST)) { // validate command
+            // Send response
+            byte[] response = new byte[DISCOVERY_RESPONSE.length()];
+            DatagramPacket responsePacket = new DatagramPacket(response, response.length, packet.getAddress(), packet.getPort());
+            getReceiveSocket().send(responsePacket);
+            logger.info("Response sent to: " + packet.getAddress() + ":" + packet.getPort());
+        } else {
+            logger.info("Error, not valid package!" + packet.getAddress() + ":" + packet.getPort());
+        }
+    }
+
+    private DatagramSocket getReceiveSocket()  {
+        try {
+            if (receiveSocket == null) {
+                receiveSocket = new DatagramSocket(8002,
+                        InetAddress.getByName("0.0.0.0")); // 0.0.0.0 for listen to all ips
+                receiveSocket.setBroadcast(true);
+            }
+        } catch (Exception e) {
+            logger.error("Broadcast Exception Encountered" + e.getMessage());
+        }
+        return receiveSocket;
+    }
+
+    private DatagramSocket getSendSocket() {
+        try {
+            if (sendSocket == null) {
+                sendSocket = new DatagramSocket(8001, InetAddress.getLocalHost());
+                sendSocket.setBroadcast(true);
+            }
+        } catch (Exception e) {
+            logger.error("Broadcast Exception Encountered" + e.getMessage());
+        }
+        return sendSocket;
+    }
+
+    public void sendBroadCast() throws Exception {
+        // Discovery request command
+        byte[] data = DISCOVERY_REQUEST.getBytes(StandardCharsets.UTF_8);
+        DatagramPacket packet = new DatagramPacket(data, data.length, InetAddress.getByName("255.255.255.255"), 8002);
+        getSendSocket().send(packet);
+        logger.info("Discovery package sent!" + packet.getAddress() + ":" + packet.getPort());
+
+        // Discovery response command
+        byte[] response = new byte[DISCOVERY_RESPONSE.length()];
+        DatagramPacket responsePacket = new DatagramPacket(response, response.length);
+        getSendSocket().receive(responsePacket);
```

Comment:
Javadoc for this method clearly says it blocks until something is received, which is what has the code stall the game :-)

Reply:
So would it be better if I try to receive something every few minutes?

Reply:
I thought something similar was going on, but I had implemented a slightly different method for sending the broadcast used here, and the broadcast was successful without the game crashing even though the recieve() method did not get anything back so when I tried listening to the broadcast using another laptop and there the game crashed when it tried to listen to the broadcast due to the recieve() method.

Reply:
I do understand its got to do something with the fact that until something is received everything else is stopped so I guess if no broadcast is received should it try after sometime?

Reply:
Would it be better to try to periodically receive the responses?

Reply:
![Screenshot (269)](https://user-images.githubusercontent.com/52852877/89729523-2caf1080-da54-11ea-8b0c-8a5bbcd5de4b.png)

As you can see here while broadcasting even though it has to receive something and is waiting the game doesnt seem to crash and it continues waiting.

Reply:
The buttons are just for testing purpose :smiley:

Reply:
So I am not exactly sure what is causing the game to crash because if it is the receive command and the socket is waiting for some response, I just tried to close the socket and retry after some time but although the processes do run in the background the game crashes.

Reply:
I'm just puzzled why there even _is_ a receive step in the first place :-)

The game server just needs to broadcast "Hey, here I am!" - I don't see what it would actually expect to ever get back. 

Client needs to be able to read the broadcast to build an entry for the server in the UI

Actually _connecting_ to the game server from a client is then a regular and unrelated process that already works fine :-)

Reply:
If I am not mistaken I remember seeing in some comment that it is supposed to broadcast every few minutes so I thought it was supposed to send a broadcast and if it doesn't receive any response then send another broadcast, or is this supposed to be for the client who instead tries to listen to a broadcast and if there isnt a broadcast try after some time?

Reply:
So while there is not broadcast either stop sending the broadcast or stop listening to the broadcast

Reply:
Or was I completely mistaken :grinning:  and is the server just supposed to send the broadcast and the client just receive the broadcast and if the client receives the broadcast add an entry in the UI?

Reply:
Now I am kind of confused what is expected in this, as in the previous PR it was attempted in this manner. So should the server just send a broadcast continuously and not stop at all and the client just try to listen to the broadcast once every 5 minutes or so and If it gets a broadcast add a UI entry to the screen?

Reply:
So should a separate UI screen be created containing information relevant to the particular network?

Reply:
So I'll repeat that I wasn't closely involved in reviewing the other PR, and I don't know if it at all made sense or was going in the right direction ... 😁 

>  So should the server just send a broadcast continuously and not stop at all and the client just try to listen to the broadcast once every 5 minutes or so and If it gets a broadcast add a UI entry to the screen?

Just thinking as a somewhat seasoned developer who has seen some things: yes, to me it would make sense to aim for that approach. Can't say that I understand why the broadcast was looking for a reply at all. But if I was trying to code in LAN game searching support then yeah, server broadcasts "Hey I'm here!" and clients look for that when the LAN games screen is open. Probably checks once on opening it, maybe again occasionally, maybe there'd be a "Refresh" button. And probably some stuff would be in secondary threads to keep things from locking.

I *think* there'd be room for a 3rd tab at the top of the current Join Game screen for "LAN"

Btw feel free to hop on Discord for more live discussion, the PRs don't always get monitored, although it helps that we have an activity channel on Discord that monitors at least _some_ things on GitHub, that's how I coincidentally noticed your comments here. On Discord you're pretty much guaranteed to get a response (eventually, thanks to time zones, hectic schedules, etc), and have a wider audience, while here some things do fall through the cracks and suddenly weeks later we're all "Wait what" and feel bad to have missed stuff :)

Reply:
Ohk sure, it just made sense for me to ask questions related to the PR in here and general questions in discord. So would questions related to the PR also be fine in discord?

Reply:
Yep questions on Discord are always fine 👍 

Actual _knowledge_ will scroll out of sight on Discord in a hurry, so it is wise to then document the outcome after any discussion on there on a relevant PR/issue. And of course it helps to have something on GitHub to point at as the starter of a discussion on Discord in the first place :-)